test/e2e/scenario/agents.scenario.ts

```diff
@@ -297,3 +305,38 @@ describe('anchor navigation', () => {
     expect(viewEvents[1].view.loading_type).toBe(ServerRumViewLoadingType.ROUTE_CHANGE)
   })
 })
+
+describe('tracing', () => {
+  it('should trace xhr', async () => {
+    const rawHeaders = await sendXhr(`${serverUrl.sameOrigin}/headers`)
+    checkRequestHeaders(rawHeaders)
+    await flushEvents()
+    await checkTraceAssociatedToRumEvent()
+  })
+
+  it('should trace fetch', async () => {
+    const rawHeaders = await sendFetch(`${serverUrl.sameOrigin}/headers`)
+    checkRequestHeaders(rawHeaders)
+    await flushEvents()
+    await checkTraceAssociatedToRumEvent()
+  })
+
+  function checkRequestHeaders(rawHeaders: string) {
+    const headers: { [key: string]: string } = JSON.parse(rawHeaders) as any
+    expect(headers['x-datadog-trace-id']).toMatch(/\d+/)
+    expect(headers['x-datadog-origin']).toBe('rum')
+  }
+
+  async function checkTraceAssociatedToRumEvent() {
+    const requests = (await waitServerRumEvents()).filter(
+      (rumEvent) =>
+        rumEvent.evt.category === 'resource' &&
+        ((rumEvent as ServerRumResourceEvent).resource.kind === 'xhr' ||
+          (rumEvent as ServerRumResourceEvent).resource.kind === 'fetch')
+    ) as ServerRumResourceEvent[]
```

Comment:
You should be able to remove all the castings with the use of `isRumResourceEvent`:

```suggestion
    const requests = (await waitServerRumEvents()).filter(
      (rumEvent) =>
        isRumResourceEvent(rumEvent) &&
        (rumEventresource.kind === 'xhr' ||
          rumEvent.resource.kind === 'fetch')
    )
```